### ISSUE TITLE
Inconsistent results of equals() for equality checking

### ISSUE THREAD

**Issue by adrianop01** [shapely/shapely#1771]
## Expected behavior and actual behavior.

For a `LineString`, after some set operations which only add points on the `LineString` itself, it can not pass the topological equality check. (Refer to code)

In the code, `pth_inbds` is a list of sub-linestring (of `pth`) after some set operations (where union of pth_inbds shall return pth topologically) and `pth` is the original linestring. While `difference()` gives empty geometry, where I would expect the two geometry to be topologically the same, the topologic equality check `equals()` gives a false value which does not make sense.

Note that `equals_exact()` is not applicable in my case as I need those extra points (from `ops.unary_union`). 

## Steps to reproduce the problem.
```
ops.linemerge(ops.unary_union(pth_inbds)).difference(pth)
><LINESTRING Z EMPTY>
```
```
ops.linemerge(ops.unary_union(pth_inbds)).difference(pth)
>False
```

The wkt values for both lines (too long to be posted here):
https://pastebin.com/clone/KN82smyZ

## Operating system

Windows 11 64 bit

## Shapely version and provenance

2.0.0 using pip

**Comment by jorisvandenbossche** [comment]
@adrianop01 the WKT output for `ops.linemerge(ops.unary_union(pth_inbds))` is truncated (there is a ", 0.3 5.91188671428...57," somewhere in the middle), so I can't try to reproduce the issue.

Could you provide the full output? Writing it to a file might be easier than printing it in a python console (which seems to truncate long strings?)

**Comment by adrianop01** [comment]
Thanks for your comment. The following two wkt representations did reproduce the same issue (`difference()` is string empty but `equals()` gives False) I mentioned:
[pth.txt](https://github.com/shapely/shapely/files/10835052/pth.txt)
[pth_modified.txt](https://github.com/shapely/shapely/files/10835053/pth_modified.txt)


```
pth_modified = to_wkt(ops.unary_union(pth_inbds),rounding_precision=21,trim=False)
f = open("pth_modified.txt", 'w')
f.writelines(pth_modified)
f.close()

pth_modified = to_wkt(pth,rounding_precision=21,trim=False)
f = open("pth.txt", 'w')
f.writelines(pth_modified)
f.close()

f = open("pth.txt",'r')
pth = from_wkt(f.read())
f.close()

f = open("pth_modified.txt",'r')
pth_modified = from_wkt(f.read())
f.close()
```
Interestingly, if I set `rounding_precision` to be something smaller, say 15, such inconsistencies do not happen:
```
pth_modified.equals(pth)
>True
pth_modified.difference(pth)
><LINESTRING Z EMPTY>
```
 Perhaps an issue of inconsistent floating point precision?

**Comment by jorisvandenbossche** [comment]
Yes, with those files I can reproduce it. Something else I observe: when checking for the difference the other way around, there _is_ some output:

```python
>>> pth.difference(pth_modified)
<MULTILINESTRING ((1.1 3.6, 1.1 3.6), (1.1 2.6, 1.1 2.6), (1.2 2.6, 1.2 2.6)...>
>>> pth_modified.difference(pth)
<LINESTRING Z EMPTY>
```

Those are a few points on the line. I suppose there are tiny floating point differences that is causing this. 

One option could be to set a precision to the coordinates (which will "round" them), and then you get the expected results:

```python
>>> pth = shapely.set_precision(pth, 1e-10)
>>> pth_modified = shapely.set_precision(pth_modified, 1e-10)
>>> pth_modified.equals(pth)
True
>>> pth_modified.difference(pth)
<LINESTRING Z EMPTY>
>>> pth.difference(pth_modified)
<LINESTRING Z EMPTY>
```
